src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,608 @@
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertLinq
+{
+    internal sealed class CSharpConvertLinqQueryToLinqMethodProvider : AbstractConvertLinqProvider
+    {
+        private static readonly TypeSyntax VarNameIdentifier = SyntaxFactory.IdentifierName("var");
+
+        protected override IAnalyzer CreateAnalyzer(SemanticModel semanticModel, CancellationToken cancellationToken)
+            => new CSharpAnalyzer(semanticModel, cancellationToken);
+
+        private sealed class CSharpAnalyzer : AnalyzerBase<QueryExpressionSyntax, SyntaxNode, SyntaxNode>
+        {
+            public CSharpAnalyzer(SemanticModel semanticModel, CancellationToken cancellationToken)
+                : base(semanticModel, cancellationToken)
+            {
+            }
+
+            protected override string Title => CSharpFeaturesResources.Convert_LINQ_query_to_foreach;
+
+            protected override bool TryConvert(QueryExpressionSyntax source, out DocumentUpdate documentUpdate)
+            {
+                if (_semanticModel.GetDiagnostics(source.Span, _cancellationToken).Any() ||
+                    source.DescendantTrivia().Any(trivia => trivia.MatchesKind(
+                        SyntaxKind.SingleLineCommentTrivia,
+                        SyntaxKind.MultiLineCommentTrivia,
+                        SyntaxKind.MultiLineDocumentationCommentTrivia) ||
+                        source.ContainsDirectives))
+                {
+                    documentUpdate = default;
+                    return false;
+                }
+
+                return TryConvertInternal(source, source.Parent, out documentUpdate);
+            }
+
+            private bool TryConvertInternal(QueryExpressionSyntax source, SyntaxNode parent, out DocumentUpdate documentUpdate)
+            {
+                switch (parent.Kind())
+                {
+                    case SyntaxKind.ReturnStatement:
+                        return TryConvertIfInReturnStatement(source, (ReturnStatementSyntax)parent, out documentUpdate);
+                    case SyntaxKind.ForEachStatement:
+                        return TryConvertIfInForEach(source, (ForEachStatementSyntax)parent, out documentUpdate);
+                    case SyntaxKind.EqualsValueClause:
+                    case SyntaxKind.SimpleAssignmentExpression:
+                        return TryReplaceWithLocalFunction(source, parent, out documentUpdate);
+                    case SyntaxKind.ParenthesizedExpression:
+                        return TryConvertInternal(source, parent.Parent, out documentUpdate);
+                    case SyntaxKind.SimpleMemberAccessExpression:
+                        return TryConvertIfInMemberAccessExpression(source, (MemberAccessExpressionSyntax)parent, out documentUpdate);
```

Comment:
can we add a test for ?.  for at least one of the tests where you have "simple.name".  Thanks!

Reply:
please see PropertyAssignmentInInvocation in tests

---
In reply to: [176297173](https://github.com/dotnet/roslyn/pull/25362#discussion_r176297173) [](ancestors = 176297173)

Reply:
No.  that test doesn't use ```?.``` .  I'd like a test to make sure we don't crash there.

Reply:
Thank you! I've added the following tests:
a?.b = query // also not sure if this one is interesting
list = query?.ToList();
Do you see more cases to cover?

---
In reply to: [176578305](https://github.com/dotnet/roslyn/pull/25362#discussion_r176578305) [](ancestors = 176578305)